README
Color Flood v0.6



debian/changelog
colorflood (0.6-1) stable; urgency=low

  * livewallpaper mode

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Sun, 29 Aug 2010 22:16:36 +0300

colorflood (0.5-1) stable; urgency=low

  * redo portrait mode layouting

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Wed, 23 Jun 2010 00:58:01 +0300

colorflood (0.4-5) stable; urgency=low

  * German translation by Falko Thomale
  * add credits

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Thu, 17 Jun 2010 20:16:37 +0300

colorflood (0.4-4) stable; urgency=low

  * replace QMessageBox with QMaemo5InformationBox (based on patch by Andrew Olmsted)

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Fri, 28 May 2010 23:25:46 +0300

colorflood (0.4-3) stable; urgency=low

  * Add rotation support
  * Fix message box rotation on game end
  * Fix help message box rotation
  * use local QMessageBox objects
  * update translation source

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Thu, 27 May 2010 17:53:50 +0300

colorflood (0.4-2) stable; urgency=low

  * debian: added Maemo-Display-Name and Maemo-Icon-26 fields, changed description

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Wed, 26 May 2010 17:51:04 +0300

colorflood (0.4-1) stable; urgency=low

  * debian: update game url
  * use libqt4 instead of libqt4-maemo5 and add bugtracker url
  * update debian/changelog
  * update debian/rules
  * fix build/run depends
  * change bugtracker url field name; fix section name
  * fix debian packaging
  * update description
  * add homepage url
  * update debian version
  * games won/played stat. Closes #4
  * rename updateTurns to updateGameState and fix games won/played when player wins with turns == maxTurns
  * fix 'game state changed' signal name
  * put labels inside vbox layout, fix their alignment. Closes #2
  * left-handed mode. Closes #5
  * fix win/lose messages and their translations
  * pretty animated flood. Closes #7
  * remove qt4_add_translation as translations are listed in resource file
  * pretty buttons with rounded corners. Closes #6

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Sun, 09 May 2010 20:20:50 +0300

colorflood (0.3-4) unstable; urgency=low

  * update description
  * add homepage url

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Mon, 26 Apr 2010 14:31:05 +0300

colorflood (0.3-3) unstable; urgency=low

  * fix debian packaging

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Mon, 26 Apr 2010 12:35:17 +0300

colorflood (0.3-2) unstable; urgency=low

  * use libqt4 instead of libqt4-maemo5 and add bugtracker url

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Mon, 26 Apr 2010 11:40:21 +0300

colorflood (0.3-1) unstable; urgency=low

  * move all resources to 'resources' directory
  * rename files and vars in the sources
  * move all files up by one level
  * use gplv3
  * update gitignore
  * show number of turns used to win the game (task 1609)
  * doxygen
  * load translation based on system locale name
  * remove extension from translation filename
  * stats: best record
  * move help button to menu bar
  * toggle fullscreen instead of exit from it
  * remove unused header
  * fix best records
  * rename bestResult to minTurnsUsedToWin to avoid misunderstanding
  * fix indentation
  * update translation
  * button: use a buttonWidthInPixels constant instead of an explicit 64
  * make long string readable by adding newlines

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Sat, 24 Apr 2010 17:38:43 +0300

colorflood (0.2-1) unstable; urgency=low

  * Help button
  * Warning fixed (in field.cpp)
  * Proper title for message boxes

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Wed, 14 Apr 2010 22:05:49 +0300

colorflood (0.1-1) unstable; urgency=low

  * Initial release

 -- Serge Ziryukin <ftrvxmtrx@gmail.com>  Wed, 14 Apr 2010 00:38:39 +0300



src/board.cpp
/*
  Copyright (C) 2010 Serge Ziryukin <ftrvxmtrx@gmail.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <QMetaType>
#include <QDataStream>
#include <QSettings>
#include <QTime>
#include <QPainter>
#include <QPaintEvent>
#include <QMessageBox>
#include <QPropertyAnimation>
#include <QMaemo5InformationBox>
#include "board.hpp"
#include "scheme.hpp"

/// board dimensions in pixels
static const int boardWidthInPixels = 420;

const int Board::boardWidthInCells[Board::NUM_SIZES] = { 14, 21, 28 };
const int Board::turnsForSize[Board::NUM_SIZES] = { 25, 35, 50 };

// we declare out QVector<Cell> metatype
// and stream operators to save whole board to settings
Q_DECLARE_METATYPE(Board::CellVector);

static QDataStream &operator<< (QDataStream &out, const Board::CellVector &rv)
{
    for (QVector<Board::Cell>::const_iterator cell = rv.begin();
         cell != rv.end();
         cell++)
    {
        out << (*cell).color;
        out << (*cell).flood;
    }

    return out;
}

static QDataStream &operator>> (QDataStream &in, Board::CellVector &rv)
{
    Board::Cell r;

    rv.clear();

    for (; !in.atEnd() ;)
    {
        in >> r.color >> r.flood;
        rv << r;
    }

    rv.pop_back();

    return in;
}

Board::Board (QWidget *parent, bool liveWallpaper, int *turns)
    : QWidget(parent),
      finished(false),
      liveWallpaper(liveWallpaper)
{
    Q_ASSERT(parent);

    setFixedSize(boardWidthInPixels, boardWidthInPixels);

    // restore board size and board itself from settings

    qRegisterMetaType<CellVector>("Board::CellVector");
    qRegisterMetaTypeStreamOperators<CellVector>("Board::CellVector");

    QSettings settings;

    int size = settings.value("board/size", SIZE_SMALL).toInt();

    if (size < SIZE_SMALL || size >= NUM_SIZES)
        size = SIZE_SMALL;

    this->size = (BoardSize)size;

    if (settings.contains("board/cells"))
        cells = settings.value("board/cells").value<CellVector>();

    this->turns = settings.value("board/turns", 0).toInt();

    if (cells.size() != boardWidthInCells[size] * boardWidthInCells[size])
        randomize();

    *turns = this->turns;

    const QVector<QColor> &scheme = Scheme::instance().getScheme();
    animatedFlood = scheme.at(cells[0].color);
    animate = new QPropertyAnimation(this, "animatedFlood", this);
    animate->setDuration(150);
}

Board::~Board ()
{
    if (!finished)
    {
        QSettings settings;

        settings.setValue("board/size", size);
        settings.setValue("board/cells", QVariant::fromValue(cells));
        settings.setValue("board/turns", turns);
    }
}

Board::BoardSize Board::getSize () const
{
    return size;
}

void Board::setSize (int size)
{
    Q_ASSERT(size >= 0 && size < NUM_SIZES);

    if (this->size == size)
        return;

    this->size = (BoardSize)size;
    randomize();
}

void Board::randomize ()
{
    Cell cell;

    cell.color = 0;
    cell.flood = false;

    cells.clear();
    cells = CellVector(boardWidthInCells[size] * boardWidthInCells[size], cell);

    qsrand(QTime(0, 0, 0).secsTo(QTime::currentTime()));

    int numColors = Scheme::instance().getScheme().size();

    for (QVector<Cell>::iterator cell = cells.begin();
         cell != cells.end();
         cell++)
    {
        (*cell).color = qrand() % numColors;
    }

    turns = 0;
    finished = false;
    emit gameStateChanged(turns, finished, false);

    // flood from top-left corner
    cells[0].flood = true;
    floodNeighbours(cells[0].color, 0, 0);

    update();
}

int Board::getNumTurnsOfSize (BoardSize size)
{
    return turnsForSize[size];
}

QColor Board::getAnimatedFlood () const
{
    return animatedFlood;
}

int Board::getCellSize (BoardSize size)
{
    return boardWidthInPixels / boardWidthInCells[size];
}

void Board::setAnimatedFlood (QColor color)
{
    animatedFlood = color;
    update();
}

void Board::tryFloodRecurse (quint8 color, int x, int y)
{
    Cell &cell = cells[x + y*boardWidthInCells[size]];

    if (!cell.flood && cell.color == color)
    {
        cell.flood = true;
        cell.fresh = true;
        floodNeighbours(color, x, y);
    }
}

void Board::floodNeighbours (quint8 color, int x, int y)
{
    int s = boardWidthInCells[size];

    cells[x + y*s].color = color;

    if (x > 0)
        tryFloodRecurse(color, x - 1, y);

    if (y > 0)
        tryFloodRecurse(color, x, y - 1);

    if (x < s - 1)
        tryFloodRecurse(color, x + 1, y);

    if (y < s - 1)
        tryFloodRecurse(color, x, y + 1);
}

void Board::paintEvent (QPaintEvent *event)
{
    QPainter painter;
    painter.begin(this);

    QRect cell = QRect(0, 0, getCellSize(size), getCellSize(size));
    const QVector<QColor> &scheme = Scheme::instance().getScheme();
    bool animating = (QAbstractAnimation::Running == animate->state());

    for (int y = 0; y < boardWidthInCells[size] ;y++)
    {
        int n = y * boardWidthInCells[size];

        for (int x = 0; x < boardWidthInCells[size] ;x++, n++)
        {
            cell.moveTo(x * cell.width(), y * cell.height());

            if (cell.intersects(event->rect()))
            {
                painter.fillRect(cell, scheme.at(cells[n].color));

                if (animating && cells[n].flood && !cells[n].fresh)
                    painter.fillRect(cell, animatedFlood);
            }
        }
    }

    painter.end();
}

void Board::flood (int colorIndex)
{
    // don't fill with the same color over and over again
    if (colorIndex == cells[0].color)
        return;

    if (finished)
        return;

    turns++;

    int oldColor = cells[0].color;

    // clear 'fresh' flag
    for (int y = 0; y < boardWidthInCells[size] ;y++)
    {
        int n = y * boardWidthInCells[size];

        for (int x = 0; x < boardWidthInCells[size] ;x++, n++)
            cells[n].fresh = false;
    }

    // flood with new color
    for (int y = 0; y < boardWidthInCells[size] ;y++)
    {
        int n = y * boardWidthInCells[size];

        for (int x = 0; x < boardWidthInCells[size] ;x++, n++)
        {
            if (cells[n].flood)
                floodNeighbours(colorIndex, x, y);
        }
    }

    const QVector<QColor> &scheme = Scheme::instance().getScheme();
    animate->setStartValue(scheme.at(oldColor));
    animate->setEndValue(scheme.at(colorIndex));
    animate->start();

    bool allFlooded = true;

    // check if all board flooded
    for (QVector<Board::Cell>::const_iterator cell = cells.begin();
         cell != cells.end();
         cell++)
    {
        if (!(*cell).flood)
        {
            allFlooded = false;
            break;
        }
    }

    QString msg;

    if (allFlooded)
    {
        finished = true;
        /*: win message */
        msg = (tr("You win!\nYou have used %1 turns out of %2.")
               .arg(turns)
               .arg(getNumTurnsOfSize(size)));
    }
    else if (getNumTurnsOfSize(size) == turns)
    {
        finished = true;
        /*: fail message */
        msg = tr("You lose!");
    }

    emit gameStateChanged(turns, finished, allFlooded);

    if (finished)
    {
        QMaemo5InformationBox::information(liveWallpaper ? NULL : this,
                                           msg,
                                           QMaemo5InformationBox::NoTimeout);
        randomize();
    }
}
